src/client/Microsoft.Identity.Client/PlatformsCommon/Shared/InMemoryPartitionedAppTokenCacheAccessor.cs

```diff
@@ -0,0 +1,210 @@
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no user accounts in a client credential flow.
+        /// </summary>
+        public MsalAccountCacheItem GetAccount(MsalAccountCacheKey accountKey)
+        {
+            throw new NotSupportedException();
+        }
+
+        public MsalAppMetadataCacheItem GetAppMetadata(MsalAppMetadataCacheKey appMetadataKey)
+        {
+            AppMetadataDictionary.TryGetValue(appMetadataKey.ToString(), out MsalAppMetadataCacheItem cacheItem);
+            return cacheItem;
+        }
+        #endregion
+
+        #region Delete
+        public void DeleteAccessToken(MsalAccessTokenCacheItem item)
+        {
+            var partitionKey = CacheKeyFactory.GetClientCredentialKey(item.ClientId, item.TenantId);
+
+            AccessTokenCacheDictionary.TryGetValue(partitionKey, out var partition);
+            if (partition == null || !partition.TryRemove(item.GetKey().ToString(), out _))
+            {
+                _logger.InfoPii(
+                    $"Cannot delete access token because it was not found in the cache. Key {item.GetKey()}.",
+                    "Cannot delete access token because it was not found in the cache.");
+            }
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no refresh tokens in a client credential flow.
+        /// </summary>
+        public void DeleteRefreshToken(MsalRefreshTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no ID tokens in a client credential flow.
+        /// </summary>
+        public void DeleteIdToken(MsalIdTokenCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+
+        /// <summary>
+        /// This method is not supported for the app token cache because
+        /// there are no user accounts in a client credential flow.
+        /// </summary>
+        public void DeleteAccount(MsalAccountCacheItem item)
+        {
+            throw new NotSupportedException();
+        }
+        #endregion
+
+        #region Get All
+
+        /// <summary>
+        /// WARNING: if partitonKey = null, this API is slow as it loads all tokens, not just from 1 partition. 
+        /// It should only to support external token caching, in the hope that the external token cache is partitioned.
+        /// </summary>
+        public virtual IReadOnlyList<MsalAccessTokenCacheItem> GetAllAccessTokens(string partitionKey = null)
+        {
+            if (string.IsNullOrEmpty(partitionKey))
+            {
+                return AccessTokenCacheDictionary.SelectMany(dict => dict.Value).Select(kv => kv.Value).ToList();
```

Comment:
Have you tried checking the performance by using TPL to collect all the items in a ConcurrentBag? Collecting data from each partition can be a separate task and it can be collected in a thread safe ConcurrentBag. This will have better perf for large number of partitions and collection. But it can be overhead for small number. The current will be good for small size. Choice depends upon the number of partitions and the size of the partitions. https://docs.microsoft.com/en-us/dotnet/api/system.collections.concurrent.concurrentbag-1?view=net-5.0

Reply:
Also there is interesting PLINQ. https://docs.microsoft.com/en-us/dotnet/standard/parallel-programming/introduction-to-plinq

Reply:
Interesting idea. We expect that with this code path, the number of elements should be relatively small. Any serious web site / web api will use external token cache serialization in conjuction with partitioning.